typedRoutes: `Link[href]`s to GET `route.ts` paths reported as problems

### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/wizardly-ride-ztcxg5?file=%2Fapp%2Fpage.tsx%3A13%2C1&workspaceId=ws_GfAuHrswXyA1DoeSwsjjjz

### To Reproduce

As in the reproduction above:

1. Install `next@15.5.0` or newer for [the `typedRoutes` feature](https://nextjs.org/blog/next-15-5#typed-routes-stable)
2. Create one `GET` [Route Handler](https://nextjs.org/docs/app/getting-started/route-handlers-and-middleware#route-handlers) with a simple path, and one with a Dynamic Segment such as [a Catch-all Segment](https://nextjs.org/docs/app/api-reference/file-conventions/dynamic-routes#catch-all-segments)
3. Link to that Route Handler using [`next/link`](https://nextjs.org/docs/app/api-reference/components/link)
4. The problems below is reported for the `href` props 💥

```bash
➜  workspace git:(main) ✗ pnpm tsc
app/page.tsx:8:13 - error TS2322: Type '"/x"' is not assignable to type 'UrlObject | RouteImpl<"/x">'.

8       <Link href="/x">/x</Link>
              ~~~~

  .next/types/link.d.ts:79:5
    79     href: __next_route_internal_types__.RouteImpl<RouteInferType> | UrlObject
           ~~~~
    The expected type comes from property 'href' which is declared here on type 'IntrinsicAttributes & LinkRestProps & { href: UrlObject | RouteImpl<"/x">; }'

app/page.tsx:9:13 - error TS2322: Type '"/y/a/b"' is not assignable to type 'UrlObject | RouteImpl<"/y/a/b">'.

9       <Link href="/y/a/b">/y/a/b</Link>
              ~~~~

  .next/types/link.d.ts:79:5
    79     href: __next_route_internal_types__.RouteImpl<RouteInferType> | UrlObject
           ~~~~
    The expected type comes from property 'href' which is declared here on type 'IntrinsicAttributes & LinkRestProps & { href: UrlObject | RouteImpl<"/y/a/b">; }'


Found 2 errors in the same file, starting at: app/page.tsx:8
```

<img width="2878" height="1704" alt="Image" src="https://github.com/user-attachments/assets/47f0cb8b-5938-4049-a7c7-84448308a3f0" />

### Current vs. Expected behavior

Current:

`typedRoutes` reports problems with links to `app/x/route.ts` and `app/y/[...ySlugs]/route.ts`

The previous implementation in `experimental.typedRoutes` did not report these problems

Expected:

`typedRoutes` reports no problems

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
  Available memory (MB): 8198
  Available CPU cores: 4
Binaries:
  Node: 20.9.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 15.5.0 // Latest available version is detected (15.5.0).
  eslint-config-next: N/A
  react: 19.1.1
  react-dom: 19.1.1
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

TypeScript

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

1. The previous implementation in `experimental.typedRoutes` did not report these problems, issue and implementation here:
   - issue: https://github.com/vercel/next.js/issues/46742
   - PR: https://github.com/vercel/next.js/pull/47185
2. The problems are not reported for `app/**/page.tsx` paths

cc @bgub

--- a/packages/next/src/server/lib/router-utils/typegen.ts
+++ b/packages/next/src/server/lib/router-utils/typegen.ts
@@ -201,18 +201,24 @@ export function generateLinkTypesFile(
   routesManifest: RouteTypesManifest
 ): string {
   // Generate serialized static and dynamic routes for the internal namespace
-  const allRoutes = {
-    ...routesManifest.appRoutes,
-    ...routesManifest.pageRoutes,
-    ...routesManifest.redirectRoutes,
-    ...routesManifest.rewriteRoutes,
-  }
+  // Build a unified set of routes across app/pages/redirect/rewrite as well as
+  // app route handlers and Pages Router API routes.
+  const allRoutesSet = new Set<string>([
+    ...Object.keys(routesManifest.appRoutes),
+    ...Object.keys(routesManifest.pageRoutes),
+    ...Object.keys(routesManifest.redirectRoutes),
+    ...Object.keys(routesManifest.rewriteRoutes),
+    // Allow linking to App Route Handlers (e.g. `/logout/route.ts`)
+    ...Object.keys(routesManifest.appRouteHandlerRoutes),
+    // Allow linking to Pages Router API routes (e.g. `/api/*`)
+    ...Array.from(routesManifest.pageApiRoutes),
+  ])
 
   const staticRouteTypes: string[] = []
   const dynamicRouteTypes: string[] = []
 
   // Process each route using the same logic as the plugin
-  for (const route of Object.keys(allRoutes)) {
+  for (const route of allRoutesSet) {
     const { isDynamic, routeType } = formatRouteToRouteType(route)
     if (isDynamic) {
       dynamicRouteTypes.push(routeType)

--- a/test/e2e/app-dir/typed-routes/app/api-test/route.ts
+++ b/test/e2e/app-dir/typed-routes/app/api-test/route.ts
@@ -1,0 +1,5 @@
+import { NextResponse } from 'next/server'
+
+export function GET() {
+  return NextResponse.json({ message: 'Hello, world!' })
+}

--- a/test/e2e/app-dir/typed-routes/typed-links.test.ts
+++ b/test/e2e/app-dir/typed-routes/typed-links.test.ts
@@ -15,6 +15,12 @@ describe('typed-links', () => {
     expect(dts).toContain(`declare module 'next/link'`)
   })
 
+  it('should include handler route from app/api-test/route.ts in generated link route definitions', async () => {
+    const dts = await next.readFile('.next/types/link.d.ts')
+    // Ensure the app route handler at app/api-test/route.ts ("/api-test") is present
+    expect(dts).toContain('`/api-test`')
+  })
+
   if (isNextStart) {
     it('should pass type checking with valid routes', async () => {
       await next.stop()

--- a/test/e2e/app-dir/typed-routes/typed-routes.test.ts
+++ b/test/e2e/app-dir/typed-routes/typed-routes.test.ts
@@ -2,7 +2,7 @@ import { nextTestSetup } from 'e2e-utils'
 
 const expectedDts = `
 type AppRoutes = "/" | "/_shop/[[...category]]" | "/dashboard" | "/dashboard/settings" | "/docs/[...slug]" | "/gallery/photo/[id]" | "/project/[slug]"
-type AppRouteHandlerRoutes = "/api/docs/[...slug]" | "/api/shop/[[...category]]" | "/api/users/[id]"
+type AppRouteHandlerRoutes = "/api-test" | "/api/docs/[...slug]" | "/api/shop/[[...category]]" | "/api/users/[id]"
 type PageRoutes = "/about" | "/users/[id]"
 type LayoutRoutes = "/" | "/dashboard"
 type RedirectRoutes = "/blog/[category]/[[...slug]]" | "/project/[slug]"

